File: resumable-change-stream.js
const { MongoClient, ReadPreference, Timestamp } = require('mongodb');

const MONGO_URL = `mongodb+srv://....mongodb.net`;
const database = `mydb`;
const collection = `foo`;
const changeStreamTokenCollection = `${collection}.changeStreamToken`;
// manually or programmatically create the change stream capped collection for storing the last read token
// db.createCollection('foo.changeStreamToken',{capped:true, size:16000, max:1})


(async () => {
    const mongoClient = await MongoClient.connect(MONGO_URL, {
        appname: 'test-change-stream',
        readPreference: ReadPreference.PRIMARY,
        useNewUrlParser: true,
    });

    const db = await mongoClient.db(database);
    setup(changeStreamTokenCollection, db);


    // I also added this to show how you can resume at a specific point in time
    const desiredDate = new Date('2023-09-30T10:00');
    const currentTimestamp = new Timestamp({ t: Math.floor(desiredDate / 1000), i: 0 })
    const resumeByTimestamp = false


    // we can use a super small and fast capped collection to store our last token processed

    const resumeByLastProcessed = true
    const resumeAfterData = await db.collection(changeStreamTokenCollection).findOne()

    // some options
    const options = {
        'fullDocument': 'updateLookup'
    };


    // use this, and create the collection changeStreamRegistry (use the name you want)
    if (resumeByLastProcessed && resumeAfterData != null) {
        options["resumeAfter"] = {
            '_data': resumeAfterData._id,
        }
    }

    // use this if you want to resume from a specific timestamp
    if (resumeByTimestamp) {
        options["startAtOperationTime"] = currentTimestamp
    }

    const changeStream = db.collection(collection).watch([
        {
            '$match': {
                'operationType': { $in: ['insert', 'update', 'delete'] }
                // and you can filter by any data from your collection
            }
        },
        // { // only project out the fields you need 
        //     '$project': {
        //         "fullDocument.a":1,
        //         "fullDocument.b":1,
        //     }
        // }
    ],
        options
    );
    console.log("Change Stream started, awaiting for events...")
    changeStream.on('change', (event) => {
        // process the event however you want
        console.log(event);

        // store the last processed event in the capped collection
        //db.collection('changeStreamRegistry').insertOne({ '_id': event._id._data })
        insertToken(event._id._data, db)
    });

})();


async function insertToken(token, db) {
    await db.collection(changeStreamTokenCollection).insertOne({ '_id': token })
}


async function setup(cappedTokenColl, db) {
    const collections = await db.listCollections().toArray();
    const collectionExists = collections.some((col) => col.name === cappedTokenColl);

    if (!collectionExists) {
        await db.createCollection(cappedTokenColl, {capped:true, size:16000, max:1});
        console.log(`Collection '${cappedTokenColl}' created.`);
    } else {
        console.log(`Collection '${cappedTokenColl}' already exists.`);
    }
}

File: resumable-change-stream.py
from pymongo import MongoClient
from bson import Timestamp
from datetime import datetime

MONGO_URL = "mongodb+srv://ilian:Password.@m0.takskf9.mongodb.net"
database_name = "mydb"
collection_name = "foo"
change_stream_token_collection = f"{collection_name}.changeStreamToken"

# Connect to MongoDB
client = MongoClient(MONGO_URL)
db = client[database_name]

def setup(capped_token_coll, db):
    collections = db.list_collection_names()
    if capped_token_coll not in collections:
        db.create_collection(capped_token_coll, capped=True, size=16000, max=1)
        print(f"Collection '{capped_token_coll}' created.")
    else:
        print(f"Collection '{capped_token_coll}' already exists.")

setup(change_stream_token_collection, db)

# Set desired date and convert it to a Timestamp
desired_date = datetime(2023, 11, 23, 15, 30, 0)
desired_timestamp = Timestamp(int(desired_date.timestamp()), 0)
resume_by_timestamp = False

# Use a small and fast capped collection to store last token processed
resume_by_last_processed = True
resume_after_data = db[change_stream_token_collection].find_one()

# Pipeline for the change stream
pipeline = [
    {
        '$match': {
            'operationType': {'$in': ['insert', 'update', 'delete']}
        }
    }
]

# Set options
options = {}

if resume_by_last_processed and resume_after_data and '_id' in resume_after_data:
    # Pass the resume token directly as a string
    options['resume_after'] = resume_after_data['_id']

if resume_by_timestamp:
    options['start_at_operation_time'] = desired_timestamp

# Create and start the change stream
change_stream = db[collection_name].watch(pipeline, **options)

print("Change Stream started, awaiting for events...")
for change in change_stream:
    print(change)
    # Store the resume token
    if '_id' in change:
        resume_token = change['_id']
        db[change_stream_token_collection].replace_one({'_id': resume_token}, {'_id': resume_token}, upsert=True)
